### PR TITLE
Created segment package

### PR DESCRIPTION
**Pull request by michalkurzeja**

### CODE REVIEW

**Comment by michalkurzeja** [comment]
This is not a complete work yet. I have created this PR to have a base for discussion.
I want to:
- discuss and improve the naming (esp. Accessor interfaces).
- discuss if the solution fulfills our needs, at least for now.

File: segment/pool.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package segment
+
+import (
+	"crypto/md5"
+	"hash"
+	"sync"
+)
+
+// hashPool represents a pool of hashes.
+type hashPool struct {
```
Comment (nrwiersma) [inline segment/pool.go]: This should be in `segment.go`

File: segment/segment.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+// Marshal returns a segment for the given identity and value.
+func Marshal(v SelfAccessor, identity Identity) (Segment, error) {
+	return MarshalWith(Wrap(v), v, identity)
+}
+
+// MarshalWith returns a segment for the given identity and value, using the specified accessor.
+func MarshalWith(accessor FieldAccessor, v interface{}, identity Identity) (Segment, error) {
+	h := pool.GetHash()
+	seg := NewSegment(identity)
```
Comment (nrwiersma) [inline segment/segment.go]: It seems inefficient to use an object here, where a string would do.

**Comment by nrwiersma** [comment]
What are the use cases you are designing for? I assume you are trying to avoid reflection.

**Comment by michalkurzeja** [comment]
FYI the code (and the discussion) has been moved to another, private repository. I'm closing this PR.

File: segment/segment.go (hunk 2 above)
Comment (michalkurzeja) [reply segment/segment.go]: The thing is, a string will not do. I want the segment to contain the data used to identify it, as well as the original values for the reverse lookup. There have been use-cases where this was needed.

**Pull request closed without merging**